ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/IdentityProviderAwareAuthService.scala

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+package com.daml.ledger.api.auth
+
+import com.auth0.jwt.JWT
+import com.daml.jwt.{JwtFromBearerHeader, JwtVerifier}
+import com.daml.jwt.domain.DecodedJwt
+import com.daml.ledger.api.domain.IdentityProviderId
+import com.daml.logging.{ContextualizedLogger, LoggingContext}
+import io.grpc.Metadata
+import spray.json._
+import com.daml.jwt.{Error => JwtError}
+
+import java.util.concurrent.{CompletableFuture, CompletionStage}
+import scala.concurrent.{ExecutionContext, Future}
+import scala.jdk.FutureConverters.FutureOps
+
+class IdentityProviderAwareAuthService(
+    defaultAuthService: AuthService,
+    configLoader: ConfigLoader,
+    jwtVerifierLoader: JwtVerifierLoader,
+)(implicit
+    executionContext: ExecutionContext,
+    loggingContext: LoggingContext,
+) extends AuthService {
+
+  private implicit val logger: ContextualizedLogger = ContextualizedLogger.get(this.getClass)
+
+  override def decodeMetadata(headers: Metadata): CompletionStage[ClaimSet] = {
+    (getAuthorizationHeader(headers) match {
+      case None => Future.successful(ClaimSet.Unauthenticated)
+      case Some(header) =>
+        parseJWTPayload(header).recover { case error =>
+          logger.warn("Authorization error: " + error.getMessage)
+          ClaimSet.Unauthenticated
+        }
+    }).asJava
+      .thenCompose(defaultAuth(headers))
+  }
+
+  def defaultAuth(
+      headers: Metadata
+  )(prevClaims: ClaimSet): CompletionStage[ClaimSet] =
+    if (prevClaims != ClaimSet.Unauthenticated)
```

Comment:
What does this code do?

Reply:
it is `orElse`-like concept for Java's `CompletionStage`. I.e. we check for the authorization IDPs and if they're giving "access denied" - we fallback to default authorization config

Reply:
Ok, so `prevClaims` are the claims resolved using the IDP aware auth service.

I'm guessing the fallback is needed because default IDP config (from the static configuration) is handled separately from the IDP configs stored in the index-db. Would it be beneficial to handle both together? E.g. by representing the default IDP as `com.daml.ledger.api.domain.IdentityProviderConfig` ?

Reply:
Default authorization might not be necessary JWKS-based, but any other config option, so it cannot be handled by this code here.

Reply:
Also Canton has it's own layer on top